src/client/auth/oidc.rs

```diff
@@ -0,0 +1,153 @@
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use bson::rawdoc;
+use futures_core::future::BoxFuture;
+use serde::Deserialize;
+
+use crate::{
+    client::{
+        auth::{
+            sasl::{SaslResponse, SaslStart},
+            AuthMechanism,
+        },
+        options::ServerApi,
+    },
+    cmap::Connection,
+    error::{Error, Result},
+};
+
+use super::{sasl::SaslContinue, Credential, MONGODB_OIDC_STR};
+
+/// The user-supplied callbacks for OIDC authentication.
+#[derive(Clone)]
+pub struct Callbacks {
+    inner: Arc<CallbacksInner>,
+}
+
+impl Callbacks {
+    /// Create a new instance with a token request callback.
+    pub fn new<F>(on_request: F) -> Self
+    where
+        F: Fn(IdpServerInfo, RequestParameters) -> BoxFuture<'static, Result<IdpServerResponse>>
+            + Send
+            + Sync
+            + 'static,
+    {
+        Self {
+            inner: Arc::new(CallbacksInner {
+                on_request: Box::new(on_request),
+            }),
+        }
+    }
+}
+
+impl std::fmt::Debug for Callbacks {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Callbacks").finish()
+    }
+}
+
+struct CallbacksInner {
+    on_request: Box<
+        dyn Fn(IdpServerInfo, RequestParameters) -> BoxFuture<'static, Result<IdpServerResponse>>
+            + Send
+            + Sync,
+    >,
+    // on_refresh: Option<Box<dyn Fn(&IdpServerInfo) -> IdpServerResponse + Send + Sync>>,
+}
+
+#[derive(Debug, Deserialize)]
+#[non_exhaustive]
+pub struct IdpServerInfo {
+    pub issuer: String,
+    #[serde(rename = "clientId")]
+    pub client_id: String,
+    #[serde(rename = "requestScopes")]
+    pub request_scopes: Option<Vec<String>>,
+}
+
+#[non_exhaustive]
+pub struct IdpServerResponse {
```

Comment:
I think we'll need to add a builder to this so that users can construct it due to non_exhaustive, ditto anything else that users will need to construct

Reply:
It's just this one, done :)